pkg/storage/ingest/reader.go

```diff
@@ -0,0 +1,331 @@
+// SPDX-License-Identifier: AGPL-3.0-only
+
+package ingest
+
+import (
+	"context"
+	"strconv"
+	"sync"
+	"time"
+
+	"github.com/go-kit/log"
+	"github.com/go-kit/log/level"
+	"github.com/grafana/dskit/backoff"
+	"github.com/grafana/dskit/multierror"
+	"github.com/grafana/dskit/services"
+	"github.com/pkg/errors"
+	"github.com/prometheus/client_golang/prometheus"
+	"github.com/prometheus/client_golang/prometheus/promauto"
+	"github.com/twmb/franz-go/pkg/kadm"
+	"github.com/twmb/franz-go/pkg/kerr"
+	"github.com/twmb/franz-go/pkg/kgo"
+	"github.com/twmb/franz-go/plugin/kprom"
+)
+
+// consumerGroup is only used to store commit offsets, not for actual consuming.
+const consumerGroup = "mimir"
+
+type record struct {
+	tenantID string
+	content  []byte
+}
+
+type recordConsumer interface {
+	consume(context.Context, []record) error
+}
+
+type PartitionReader struct {
+	services.Service
+
+	kafkaAddress  string
+	kafkaTopic    string
+	kafkaClientID string
+	partitionID   int32
+
+	client    *kgo.Client
+	admClient *kadm.Client
+
+	consumer recordConsumer
+	metrics  readerMetrics
+
+	commitInterval time.Duration
+	commitLoopWg   *sync.WaitGroup
+	commitFetches  chan kgo.Fetches
+
+	logger log.Logger
+}
+
+func NewReaderForPusher(kafkaAddress, kafkaTopic, kafkaClientID string, partitionID int32, pusher Pusher, logger log.Logger, reg prometheus.Registerer) (*PartitionReader, error) {
+	metrics := newReaderMetrics(partitionID, reg)
+	consumer := newPusherConsumer(pusher, metrics, logger)
+	return newReader(kafkaAddress, kafkaTopic, kafkaClientID, partitionID, consumer, logger, metrics)
+}
+
+func newReader(kafkaAddress, kafkaTopic, kafkaClientID string, partitionID int32, consumer recordConsumer, logger log.Logger, metrics readerMetrics) (*PartitionReader, error) {
+	r := &PartitionReader{
+		kafkaAddress:   kafkaAddress,
+		kafkaTopic:     kafkaTopic,
+		kafkaClientID:  kafkaClientID,
+		partitionID:    partitionID,
+		consumer:       consumer, // TODO consume records in parallel
+		commitInterval: time.Second,
+		commitLoopWg:   &sync.WaitGroup{},
+		commitFetches:  make(chan kgo.Fetches, 1),
+		metrics:        metrics,
+		logger:         log.With(logger, "partition", partitionID),
+	}
+
+	r.Service = services.NewBasicService(r.start, r.run, r.stop)
+	return r, nil
+}
+
+func (r *PartitionReader) start(ctx context.Context) error {
+	offset, err := r.fetchLastCommittedOffsetWithRetries(ctx)
```

Comment:
Any specific reason why `fetchLastCommittedOffset()` creates its own client, instead of moving this call after the creation of the client below and then just using it?

Reply:
I can potentially change SetOffsets to allow this, it's a pretty old function and I've removed some internal restrictions that made SetOffsets more limited in the past. Not sure on the timeline of that work though (this isn't the first need to improve SetOffsets that I've seen in the past few months)

Reply:
Thanks for taking a look :) It will be useful if we can simplify the code, but it's more of a nice-to-have, than a requirement at this point.